### PR TITLE
 feat: add anthropic-messages protocol for native Anthropic Messages API support    Fixes #269

### LINKED ISSUE #269: Anthropic/Claude API direct integration fails with 404 error
## Description

The Anthropic/Claude direct API integration in picoclaw fails with a 404 error when using a valid API key and model name, despite the API key working correctly when tested directly with curl.

## Environment
- **picoclaw version**: Latest (commit 341dbd3)
- **Platform**: Docker (alpine:3.23, golang:1.26.0-alpine)
- **Config**:
  - Provider: `anthropic`
  - Model: `claude-opus-4-6` (verified valid via Anthropic API)
  - API Key: Valid (tested successfully with curl)

## Steps to Reproduce

1. Configure `config.json` with Anthropic provider:
```json
{
  "agents": {
    "defaults": {
      "provider": "anthropic",
      "model": "claude-opus-4-6"
    }
  },
  "providers": {
    "anthropic": {
      "api_key": "<valid-api-key>",
      "api_base": "https://api.anthropic.com"
    }
  }
}
```

2. Run: `docker compose run --rm picoclaw-agent -m "Hello"`

## Expected Behavior
Picoclaw should use the native `ClaudeProvider` to successfully communicate with Anthropic's API and return a response.

## Actual Behavior
Error: `LLM call failed: API request failed: Status: 404, Body: ` (empty body)

## Verification
The same API key and model work perfectly when tested directly:
```bash
curl -X POST https://api.anthropic.com/v1/messages \
  -H "x-api-key: <api-key>" \
  -H "anthropic-version: 2023-06-01" \
  -H "content-type: application/json" \
  -d '{"model": "claude-opus-4-6", "max_tokens": 100, "messages": [{"role": "user", "content": "Hello"}]}'
```

**Result**: Success - receives proper response from Claude Opus 4.6

## Root Cause
The issue is in `pkg/providers/http_provider.go` in the `CreateProvider` function. Even when `provider` is set to `"anthropic"` and there's a valid API key, the code falls through to use `HTTPProvider` instead of the native `ClaudeProvider`.

**Two problematic code paths:**

1. **Lines 255-265**: Explicit provider selection
2. **Lines 363-372**: Model name-based selection

Both paths set API configuration for `HTTPProvider` (which uses OpenAI-compatible `/chat/completions` endpoint), but Anthropic's native API has different endpoint structure and requires the proper SDK usage via `ClaudeProvider`.

## Proposed Fix

Modify both code paths to return `NewClaudeProvider` directly when an Anthropic API key is present:

**Location 1** (line 255-265):
```go
case "anthropic", "claude":
    if cfg.Providers.Anthropic.APIKey != "" || cfg.Providers.Anthropic.AuthMethod != "" {
        if cfg.Providers.Anthropic.AuthMethod == "oauth" || cfg.Providers.Anthropic.AuthMethod == "token" {
            return createClaudeAuthProvider()
        }
        // FIX: Use native ClaudeProvider when API key is provided
        if cfg.Providers.Anthropic.APIKey != "" {
            return NewClaudeProvider(cfg.Providers.Anthropic.APIKey), nil
        }
        // ... rest of fallback code
    }
```

**Location 2** (line 363-372):
```go
case (strings.Contains(lowerModel, "claude") || strings.HasPrefix(model, "anthropic/")) && (cfg.Providers.Anthropic.APIKey != "" || cfg.Providers.Anthropic.AuthMethod != ""):
    if cfg.Providers.Anthropic.AuthMethod == "oauth" || cfg.Providers.Anthropic.AuthMethod == "token" {
        return createClaudeAuthProvider()
    }
    // FIX: Use native ClaudeProvider when API key is provided
    if cfg.Providers.Anthropic.APIKey != "" {
        return NewClaudeProvider(cfg.Providers.Anthropic.APIKey), nil
    }
    // ... rest of fallback code
```

## Additional Context
- The `ClaudeProvider` exists and works correctly (uses official Anthropic SDK)
- The issue prevents users from using Claude/Anthropic directly with their API keys
- Users are forced to use OpenRouter or other proxies as a workaround

## Workarounds
Currently, users must use OpenRouter or other proxy services to access Claude models, which adds unnecessary complexity and costs.

### PR DESCRIPTION
**Pull request by hyperwd**
## 概述
  添加 `anthropic-messages` 协议支持，与 [openclaw](https://github.com/openclaw/openclaw) 的配置模式保持兼容，解决 issue #269 中的 404 错误。

  ## 设计参考
  - **配置格式**: 与 openclaw 的 `api: "anthropic-messages"` 保持一致
  - **请求格式**: 遵循 Anthropic Messages API 官方规范
  - **使用场景**: 支持 MiniMax、Xiaomi、Synthetic 等第三方代理
  - **协议标识**: `anthropic-messages/{model}`

  ## 问题 (Issue #269)
  使用 Anthropic API 时返回 404 错误。现有实现使用 OpenAI 兼容格式 (`/v1/chat/completions`)，而 Anthropic 原生 API 使用 `/v1/messages` 端点。

  ## 配置示例
  ### 标准 Anthropic API
  ```yaml
  model: "anthropic-messages/claude-opus-4-6"
  api_base: "https://api.anthropic.com"
  api_key: "sk-ant-..."

  第三方代理（参考 openclaw MiniMask/Synthetic 配置）

  # openclaw 兼容配置
  providers:
    minimax:
      baseUrl: "https://api.minimax.io/anthropic"
      api: "anthropic-messages"
      apiKey: "${MINIMAX_API_KEY}"

  # picoclaw 等效配置
  model: "anthropic-messages/MiniMax-M2.5"
  api_base: "https://api.minimax.io/anthropic"
  api_key: "sk-..."

  文档更新

  - ✅ config/config.example.json - 添加 anthropic-messages 示例
  - ✅ README.md - 新增 "Anthropic Messages API" 章节（参考 openclaw 文档）
  - ✅ README.zh.md - 中文文档说明

  测试

  - ✅ 单元测试：64.2% 覆盖率
  - ✅ 实际 API：成功验证（使用第三方代理端点）
  - ✅ 编译：无错误
  - ✅ 兼容性：不影响现有 anthropic protocol
  ```

## 🗣️ Type of Change
- [x] 🐞 Bug fix (non-breaking change which fixes an issue)
- [x] ✨ New feature (non-breaking change which adds functionality)
- [x] 📖 Documentation update
- [ ] ⚡ Code refactoring (no functional changes, no api changes)

## 🤖 AI Code Generation
- [ ] 🤖 Fully AI-generated (100% AI, 0% Human)
- [x] 🛠️ Mostly AI-generated (AI draft, Human verified/modified)
- [ ] 👨‍💻 Mostly Human-written (Human lead, AI assisted or none)

## 🔗 Related Issue

Fixes #269 
## 📚 Technical Context (Skip for Docs)
- **Reference URL:**
- **Reasoning:**

## 🧪 Test Environment
- **Hardware:** 
- **OS:** 
- **Model/Provider:** 
- **Channels:** 

## 📸 Evidence (Optional)
<details>
<summary>Click to view Logs/Screenshots</summary>

</details>

## ☑️ Checklist
- [x] My code/docs follow the style of this project.
- [x] I have performed a self-review of my own changes.
- [x] I have updated the documentation accordingly.

### CODE REVIEW

File: pkg/providers/anthropic_messages/provider.go [hunk 1]
```diff
@@ -0,0 +1,392 @@
...
+	UsageInfo              = protocoltypes.UsageInfo
+	Message                = protocoltypes.Message
+	ToolDefinition         = protocoltypes.ToolDefinition
+	ToolFunctionDefinition = protocoltypes.ToolFunctionDefinition
+)
+
+const (
+	defaultAPIVersion   = "2023-06-01"
+	defaultBaseURL      = "https://api.anthropic.com/v1"
+	defaultRequestTimeout = 120 * time.Second
```
Comment (huaaudio) [inline pkg/providers/anthropic_messages/provider.go:35]: 120s might be too long for this case, I'd suggest 20 seconds to be already good enough

File: pkg/providers/anthropic_messages/provider.go [hunk 2]
```diff
@@ -0,0 +1,392 @@
...
+// buildRequestBody converts internal message format to Anthropic Messages API format.
+func buildRequestBody(
+	messages []Message,
+	tools []ToolDefinition,
+	model string,
+	options map[string]any,
+) (map[string]any, error) {
+	result := map[string]any{
+		"model":     model,
+		"max_tokens": int64(4096),
```
Comment (huaaudio) [inline pkg/providers/anthropic_messages/provider.go]: 4096 hardcoded here would not align well with the config. Try to set that to match `max_tokens` in `config.json` .

File: pkg/providers/anthropic_messages/provider.go [hunk 3]
```diff
@@ -0,0 +1,392 @@
...
+	ToolFunctionDefinition = protocoltypes.ToolFunctionDefinition
+)
+
+const (
+	defaultAPIVersion   = "2023-06-01"
+	defaultBaseURL      = "https://api.anthropic.com/v1"
+	defaultRequestTimeout = 120 * time.Second
+)
+
+// Provider implements Anthropic Messages API via HTTP (without SDK).
```
Comment (huaaudio) [inline pkg/providers/anthropic_messages/provider.go:38]: Unlike the HTTP provider (which may have retries), this implementation makes a single HTTP call. Consider adding retry logic for transient failures, especially since this is a network-based provider.

**Comment by hyperwd** [comment]
感谢 @huaaudio 的审查！关于这些建议：

  ## 1. 超时时间 120s

  我检查了项目内其他 provider 的实现，发现都是统一的 120s：

  - `pkg/providers/openai_compat/provider.go:41` - `defaultRequestTimeout = 120 * time.Second`
  - `pkg/providers/http_provider.go:41` - 也是 120s

  LLM API 在处理复杂任务（工具调用、长文本生成）时响应时间可能较长，20s 容易导致超时。为了保持一致性，建议维持 120s。

  ## 2. max_tokens 默认值 4096

  这个默认值是必要的，因为 **Anthropic Messages API 要求 `max_tokens` 为必填参数**（与 OpenAI 不同，不传会报错）。

  代码已支持从 `options["max_tokens"]` 动态覆盖（见 provider.go:154-155 行），实现方式与 `openai_compat/provider.go:128` 一致。

  对比：
  | Provider | 是否需要默认 max_tokens | 原因 |
  |----------|----------------------|------|
  | openai_compat | ❌ 不需要 | OpenAI API 的 max_tokens 是可选参数 |
  | anthropic_messages | ✅ 必须设置 | Anthropic Messages API 的 max_tokens 是**必填**参数 |

  ## 3. 重试逻辑

  我检查了 `http_provider` 和 `openai_compat provider` 的实现（见 openai_compat/provider.go:181），目前都直接调用 `httpClient.Do()`，**没有重试机制**。

  重试逻辑建议后续统一为所有 provider 添加，不应作为本 PR 的阻塞项。

  ## Linter 错误修复

  ✅ **已修复所有 linter 错误**：
  - commit 2572500: gofmt 格式问题
  - commit c748a86: HTTP header canonical form + 其他格式问题

  请帮忙重新审查，谢谢！

**Comment by hyperwd** [comment]
@yinwm 感谢您的详细审查！关于这四个问题：

  ---
  ✅ 问题 1：normalizeBaseURL 边界情况 Bug

  已修复。使用 strings.CutSuffix 正确处理包含 /v1 的 URL（如 https://api.example.com/v1/proxy），避免重复添加 /v1 后缀。

  // 修复前：https://api.example.com/v1/proxy → https://api.example.com/v1/proxy/v1 ❌
  // 修复后：https://api.example.com/v1/proxy → https://api.example.com/v1/proxy ✅

  参考了 pkg/providers/anthropic/provider.go:385-400 的实现方式。

  ---
  ✅ 问题 2：死代码（apiBase 空检查）

  已修复。删除了 provider.go:76-78 的空检查代码，因为 normalizeBaseURL() 已将空字符串转为默认 URL，此检查永远不会触发。

  ---
  问题 3：max_tokens 硬编码

  与现有代码一致：pkg/providers/anthropic/provider.go:207-210 使用了完全相同的模式（硬编码 4096 + options 覆盖）。

  API 要求：Anthropic Messages API 的 max_tokens 是必填参数（与 OpenAI 不同），不传会返回 400 错误，必须提供默认值。

  支持配置覆盖：代码已支持从 options["max_tokens"] 动态覆盖（provider.go:154-156），与 openai_compat provider 实现一致。

  ---
  问题 4：架构设计

  我理解代码重复和维护负担的担忧，但这里有一个关键背景：

  问题根源

  现有的 "anthropic" 协议（factory_provider.go:134-140）使用 OpenAI 兼容格式（/chat/completions），而 Issue #269 中的第三方代理（MiniMax/Synthetic）只支持 Anthropic 原生格式（/messages），导致 404 错误。

  格式差异

  - OpenAI 兼容格式（anthropic + API key）：{"model": "...", "messages": [...], ...} → /chat/completions
  - Anthropic 原生格式（anthropic-messages）：{"model": "...", "max_tokens": 4096, "messages": [...]} → /messages

  两种格式的请求体结构不同，无法通过简单配置切换。本 PR 创建的 anthropic-messages provider 专门处理原生格式。

  是否可以扩展

  如果要在现有 anthropic.Provider（SDK-based）中支持，需要同时处理两种格式，会增加复杂度。创建独立的 provider 更符合单一职责原则。

**Comment by hyperwd** [comment]
@yinwm 感谢详细的代码审查！

  关于 问题 3: 硬编码 max_tokens 值，我已修复此问题：

  ✅ 已修复

  Commit: 94c7d87 - fix(providers): remove hardcoded max_tokens in anthropic-messages provider

  修改内容

  之前（硬编码 4096）：
  result := map[string]any{
      "model":      model,
      "max_tokens": int64(4096),  // ❌ 硬编码
      "messages":   []any{},
  }

  现在（直接使用 options）：
  // max_tokens is required and guaranteed by agent loop
  maxTokens, ok := asInt(options["max_tokens"])
  if !ok {
      return nil, fmt.Errorf("max_tokens is required in options")
  }

  result := map[string]any{
      "model":      model,
      "max_tokens": int64(maxTokens),  // ✅ 从 config 读取
      "messages":   []any{},
  }

  效果

  现在 provider 正确遵循配置系统的默认值：
  - Config 默认值: 32768 (pkg/config/defaults.go)
  - 系统兜底值: 8192 (pkg/agent/instance.go:118-121)
  - 数据流: Config → Agent Instance → Agent Loop → Provider ✅

  测试验证

  - ✅ 所有单元测试通过（包括新增的错误处理测试）
  - ✅ golangci-lint: 0 issues
  - ✅ gofmt 格式检查通过

**Comment by hyperwd** [comment]
@yinwm 感谢您的详细审查！我已经修复了所有指出的问题：

---

✅ 问题 1：ToolCalls nil vs empty slice（必须修复）
- 已将 `var toolCalls []ToolCall` 改为 `toolCalls := make([]ToolCall, 0)`
- 确保 JSON 序列化返回 `[]` 而不是 `null`
- 详见 commit a22e027

---

✅ 问题 2：错误处理细化（建议）
- 已添加对常见 HTTP 状态码的详细错误信息
- 包括 401（认证失败）、429（限流）、400（错误请求）、404、500、503
- 便于用户快速识别和排查问题

---

✅ 问题 3：边界测试（认可建议）
- 新增 `TestBuildRequestBodyEdgeCases`：空消息列表、长系统消息、多系统消息合并等
- 新增 `TestParseResponseBodyEdgeCases`：空内容块、多工具调用、格式错误等
- 所有测试通过（包括新增的 7 个测试用例）

---

✅ 问题 5：文档优化（认可建议）
- 更新了 README.md 和 README.zh.md
- 明确说明 `anthropic` 协议使用 OpenAI 兼容格式（`/v1/chat/completions`）
- 明确说明 `anthropic-messages` 协议使用 Anthropic 原生格式（`/v1/messages`）
- 列出第三方代理的具体场景（MiniMax、Synthetic 等）

---

所有 CI 检查已通过（golangci-lint: 0 issues, go test: 全部通过），请帮忙重新审查，谢谢！

**Review by yinwm — APPROVED** [review]
## Code Review Summary

This PR has been thoroughly reviewed and all previously raised issues have been addressed:

### ✅ Fixed Issues
1. **normalizeBaseURL edge case** - Fixed with `strings.CutSuffix`
2. **Dead code (apiBase check)** - Removed
3. **Hardcoded max_tokens** - Now reads from options
4. **ToolCalls nil vs empty slice** - Fixed with `make([]ToolCall, 0)`
5. **Error handling** - Now distinguishes HTTP status codes (401/429/400/404/500/503)
6. **Edge case tests** - Added comprehensive test coverage

### Code Quality
- Clean code structure following project patterns
- Good test coverage (64.2%)
- Complete documentation in both English and Chinese
- Responsive author who addressed all feedback across 7 iterations

### Minor Notes (Non-blocking)
- Test could include `Arguments` field validation in `TestParseResponseBody`
- `normalizeBaseURL` may produce double `/v1` for URLs like `https://api.example.com/v1/proxy` (rare edge case, likely user misconfiguration)

**Recommendation: ✅ Approve for merge**

Great work @hyperwd! 🎉

**Comment by Orgmar** [comment]
@hyperwd anthropic-messages协议支持做得很扎实，和openclaw的配置模式保持兼容，还顺便解决了issue #269的404问题。对MiniMax这些第三方代理的支持也很实用。

对了，我们有个 **PicoClaw Dev Group**，在Discord上方便贡献者们直接交流。感兴趣的话，给 `support@sipeed.com` 发一封邮件，主题写 `[Join PicoClaw Dev Group] + 你的GitHub账号`，我们会发送Discord邀请链接！
